src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultGeneratedDocumentPublisher.cs

```diff
@@ -123,6 +124,16 @@ public override void PublishCSharp(ProjectKey projectKey, string filePath, Sourc
             HostDocumentVersion = hostDocumentVersion,
         };
 
+        // HACK: We know about a document being in multiple projects, but despite having ProjectKeyId in the request, currently the other end
+        // of this LSP message only knows about a single document file path. To prevent confusing them, we just send an update for the first project
+        // in the list.
+        if (_projectSnapshotManager is not null &&
+            _projectSnapshotManager.GetLoadedProject(projectKey) is { } projectSnapshot &&
+            _projectSnapshotManager.GetAllProjectKeys(projectSnapshot.FilePath).First() != projectKey)
```

Comment:
Can `GetAllProjectKeys(...)` return an empty array?

Reply:
Not without the world being upside down. We're asking for a project from the manager, and then for a list of all projects that have that projects file path. If the project we got back doesn't have its own file path, I don't know what to think :)

Reply:
Actually, maybe I'm thinking about this wrong. I guess without dispatcher threads, its entirely possible the project could be removed between the GetLoadedProject and GetAllProjectKeys call. Either all of this code needs to be a lot more defensive, or we need more specific helpers in the snapshot manager that can run things under its locks.

Thoughts @ryzngard?

I think I'll leave this for now, since its a relatively temporary mitigation for an issue, but I guess long term I need to re-adjust how I approach these things.

Reply:
Right now I wouldn't expect it to be an issue. The only thing we've completely moved off the dispatcher thread is read operations. Longterm I think we should really have some sort of immutable encapsulation of the current project states (similar to a `Solution` ;) )

One easy fix if we get worried is to add a `GetAllProjectKeys(projectKey)` method. This seems like mitigation right now though and won't hit a problem until something else changes.